File: user-service/src/main/java/io/ecommerce/userservice/error/ErrorDetail.java
package io.ecommerce.userservice.error;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author : choi-ys
 * @date : 2022/01/07 6:43 오후
 */
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ErrorDetail {
    private String object;
    private String field;
    private String code;
    private String rejectMessage;
    private Object rejectedValue;

    private ErrorDetail(
            final String object,
            final String field,
            final String code,
            final String rejectMessage,
            final Object rejectedValue
    ) {
        this.object = object;
        this.field = field;
        this.code = code;
        this.rejectMessage = rejectMessage;
        this.rejectedValue = rejectedValue;
    }

    public static List<ErrorDetail> of(final BindingResult bindingResult) {
        List<FieldError> fieldErrors = bindingResult.getFieldErrors();
        return fieldErrors.stream()
                .map(it ->
                        new ErrorDetail(
                                it.getObjectName(),
                                it.getField(),
                                it.getCode(),
                                it.getDefaultMessage(),
                                it.getRejectedValue()
                        )
                ).collect(Collectors.toList());
    }
}

File: product-service/src/main/kotlin/io/ecommerce/productgservice/error/ErrorDetail.kt
package io.ecommerce.productgservice.error

import org.springframework.validation.BindingResult
import org.springframework.validation.FieldError
import java.util.stream.Collectors

/**
 * @author : choi-ys
 * @date : 2022-01-21 오전 12:41
 */
data class ErrorDetail private constructor(
    val `object`: String,
    val field: String,
    val code: String,
    val rejectMessage: String,
    val rejectedValue: Any,
) {
    companion object {
        fun of(bindingResult: BindingResult): List<ErrorDetail> {
            val fieldErrors = bindingResult.fieldErrors
            return fieldErrors.stream()
                .map { it: FieldError ->
                    ErrorDetail(
                        it.objectName,
                        it.field,
                        it.code,
                        it.defaultMessage,
                        it.rejectedValue
                    )
                }.collect(Collectors.toList())
        }
    }
}